alf/algorithms/hypernetwork_algorithm.py

```diff
@@ -0,0 +1,430 @@
+            num_particles (int): size of sampled ensemble.
+            state: not used.
+
+        Returns:
+            AlgorithmStep: outputs with shape (batch_size, output_dim)
+        """
+        if params is None:
+            params = self.sample_parameters(num_particles=num_particles)
+        self._param_net.set_parameters(params)
+        outputs, _ = self._param_net(inputs)
+        return AlgStep(output=outputs, state=(), info=())
+
+    def train_iter(self, num_particles=None, state=None):
+        """Perform one epoch (iteration) of training.
+        
+        Args:
+            num_particles (int): number of sampled particles.
+            state: not used
+
+        Return:
+            mini_batch number
+        """
+
+        assert self._train_loader is not None, "Must set data_loader first."
+        alf.summary.increment_global_counter()
+        with record_time("time/train"):
+            loss = 0.
+            if self._loss_type == 'classification':
+                avg_acc = []
+            for batch_idx, (data, target) in enumerate(self._train_loader):
+                data = data.to(alf.get_default_device())
+                target = target.to(alf.get_default_device())
+                alg_step = self.train_step((data, target),
+                                           num_particles=num_particles,
+                                           state=state)
+                loss_info, params = self.update_with_gradient(alg_step.info)
+                loss += loss_info.extra.generator.loss
+                if self._loss_type == 'classification':
+                    avg_acc.append(alg_step.info.extra.generator.extra)
+        acc = None
+        if self._loss_type == 'classification':
+            acc = torch.as_tensor(avg_acc).mean() * 100
+        if self._logging_training:
+            if self._loss_type == 'classification':
+                logging.info("Avg acc: {}".format(acc))
+            logging.info("Cum loss: {}".format(loss))
+        self.summarize_train(loss_info, params, cum_loss=loss, avg_acc=acc)
+
+        return batch_idx + 1
+
+    def train_step(self,
+                   inputs,
+                   num_particles=None,
+                   entropy_regularization=None,
+                   state=None):
+        """Perform one batch of training computation.
+
+        Args:
+            inputs (nested Tensor): input training data. 
+            num_particles (int): number of sampled particles. 
+            state: not used
+
+        Returns:
+            AlgorithmStep:
+                outputs: Tensor with shape (batch_size, dim)
+                info: LossInfo
+        """
+        params = self.sample_parameters(num_particles=num_particles)
+        if entropy_regularization is None:
+            entropy_regularization = self._entropy_regularization
+
+        return self._generator.train_step(
+            inputs=None,
+            loss_func=functools.partial(neglogprob, inputs, self._param_net,
```

Comment:
If neglogprob is a member function, this can be simplified as `partial(neglogprob, inputs)`

Reply:
updated